src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "not a function pointer: " + id2string(restriction.first),
+        "--restrict-function-pointer"};
+    }
+    auto const &function_type =
+      to_pointer_type(function_pointer_type).subtype();
+    if(function_type.id() != ID_code)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "not a function pointer: " + id2string(restriction.first),
+        "--restrict-function-pointer"};
+    }
+    auto const &ns = namespacet{goto_model.symbol_table};
+    for(auto const &function_pointer_target : restriction.second)
+    {
+      auto const function_pointer_target_sym =
+        goto_model.symbol_table.lookup(function_pointer_target);
+      if(function_pointer_target_sym == nullptr)
+      {
+        throw invalid_command_line_argument_exceptiont{
+          "symbol not found: " + id2string(function_pointer_target),
+          "--restrict-function-pointer"};
+      }
+      auto const &function_pointer_target_type =
+        function_pointer_target_sym->type;
+      if(function_type != function_pointer_target_type)
+      {
+        throw invalid_command_line_argument_exceptiont{
+          "type mismatch: `" + id2string(restriction.first) + "' points to `" +
+            type2c(function_type, ns) + "', but restriction `" +
+            id2string(function_pointer_target) + "' has type `" +
+            type2c(function_pointer_target_type, ns) + "'",
+          "--restrict-function-pointer"};
+      }
+    }
+  }
+}
+
+source_locationt make_function_pointer_restriction_assertion_source_location(
+  source_locationt source_location,
+  const function_pointer_restrictionst::value_type restriction)
+{
+  std::stringstream comment;
+  comment << "dereferenced function pointer at " << restriction.first
+          << " must be ";
+  if(restriction.second.size() == 1)
+  {
+    comment << *restriction.second.begin();
+  }
+  else
+  {
+    comment << "one of [";
+    bool first = true;
+    for(auto const &target : restriction.second)
+    {
+      if(!first)
+      {
+        comment << ", ";
+      }
+      else
+      {
+        first = false;
+      }
+      comment << target;
+    }
+    comment << ']';
+  }
+  source_location.set_comment(string2id(comment.str()));
+  source_location.set_property_class(ID_assertion);
+  return source_location;
+}
+
+template <typename Handler>
+void for_each_function_call(goto_functiont &goto_function, Handler handler)
+{
+  for_each_goto_location_if(
+    goto_function,
+    [](goto_programt::targett target) { return target->is_function_call(); },
+    handler);
+}
+} // namespace
+
+void restrict_function_pointers(
+  goto_modelt &goto_model,
+  const function_pointer_restrictionst &restrictions)
+{
+  typecheck_function_pointer_restrictions(goto_model, restrictions);
+
+  for(auto &goto_function : goto_model.goto_functions.function_map)
+  {
+    auto &goto_function_body = goto_function.second.body;
```

Comment:
This gets pretty indented - consider breakin gout the body of the loop into a function that deals with bodies to save one level of indentiation

Reply:
Moved the body of the if into a separate function